lib/bq_stream/archive.rb

```diff
@@ -0,0 +1,136 @@
+      buffer = []
+
+      # Set all BigQuery variables
+      client_id =     ENV['CLIENT_ID']
+      service_email = ENV['SERVICE_EMAIL']
+      key =           ENV['KEY']
+      project_id =    ENV['PROJECT_ID']
+      bq_table_name = ENV['BQ_TABLE_NAME'] || 'bq_datastream'
+      time_zone =     ENV['TIMEZONE'] || 'UTC'
+
+      # Create BigQuery client connection
+      require 'big_query'
+      opts = {}
+      opts['client_id']     = client_id
+      opts['service_email'] = service_email
+      opts['key']           = key
+      opts['project_id']    = project_id
+      opts['dataset']       = dataset
+      @bq_archiver = BigQuery::Client.new(opts)
+
+      # Create dataset if not present in BigQuery
+      @bq_archiver.create_dataset(dataset) unless @bq_archiver.datasets_formatted.include?(dataset)
+
+      # Create table in dataset if not present in BigQuery
+      unless @bq_archiver.tables_formatted.include?(bq_table_name)
+        bq_table_schema = { table_name:   { type: 'STRING', mode: 'REQUIRED' },
+                            record_id:    { type: 'INTEGER', mode: 'REQUIRED' },
+                            attr:         { type: 'STRING', mode: 'NULLABLE' },
+                            new_value:    { type: 'STRING', mode: 'NULLABLE' },
+                            updated_at:   { type: 'TIMESTAMP', mode: 'REQUIRED' } }
+        @bq_archiver.create_table(bq_table_name, bq_table_schema)
+      end
+
+      BqStream.log(:info, "#{Time.now}: ***** Start Update Oldest Record Rows *****")
+
+      # Check to make sure all attributes in BigQuery are represented in OldestRecord table
+      old_records = @bq_archiver.query('SELECT table_name, attr FROM '\
+                                      "[#{project_id}:#{dataset}.#{bq_table_name}] "\
+                                      'GROUP BY table_name, attr')
+      if old_records['rows']
+        old_records['rows'].each do |r|
+          table = r['f'][0]['v']
+          trait = r['f'][1]['v']
+          unless trait.nil?
+            BqStream::OldestRecord.find_or_create_by(table_name: table, attr: trait)
+          end
+        end
+      end
+
+      # Reset all rows archived status to false to run through all tables
+      BqStream::OldestRecord.where.not(table_name: '! revision !').update_all(archived: false)
+      BqStream.log(:info, "#{Time.now}: ***** End Update Oldest Record Rows *****")
+
+      # Stop processing when all tables are archived
+      until BqStream::OldestRecord.where(archived: false).empty? && buffer.empty?
+        BqStream::OldestRecord.table_names.each do |table|
+          table_class = table.constantize
+
+          BqStream.log(:info, "#{Time.now}: ***** Cycle Table: #{table} *****")
+          oldest_attr_recs = BqStream::OldestRecord.where(table_name: table)
+
+          # See if there are any records for given table in BigQuery
+          record_id_query = @bq_archiver.query('SELECT table_name, attr, min(record_id) as earliest_record_id '\
```

Comment:
This isn't a problem for us, but something to consider: A lot of databases now use UUIDs as the primary key (what IW uses alongside a normal ID). It's a random hex number and not assigned in ascending order, so this would not work for that. We don't use that so this is nothing to change right now, but something to consider in the future.